[14.0][IMP] hr_timesheet_purchase_order: create PO automatically

dessanhemrayev opened the pull request:
Create PO automatically

dessanhemrayev committed:
[IMP] hr_timesheet_purchase_order: create PO automatically

[FIX]hr_timesheet_purchase_order:change field type

Change field type `selection` to `integer`
Add new method border min max and chech month `february`
Fix and update tests

[FIX] hr_timesheet_purchase_order: add method

geomer198 commented on hr_timesheet_purchase_order/models/hr_employee_base.py:
```diff
@@ -8,3 +12,237 @@ class HrEmployeeBase(models.AbstractModel):
     billing_partner_id = fields.Many2one("res.partner")
+
+    is_auto_po_generate = fields.Boolean(
+        string="Automatic PO generation", default=False
```
By default boolean field set "false" value.

dessanhemrayev replied:
ok,I'll fix it

geomer198 commented on hr_timesheet_purchase_order/models/hr_employee_base.py:
```diff
@@ -8,3 +12,237 @@ class HrEmployeeBase(models.AbstractModel):
+    recurrence_id = fields.Many2one("hr.timesheet.recurrence", copy=False)
+
+    is_send_po = fields.Boolean(
+        string="Send RFQ by email after creation", default=False
```
Here too

dessanhemrayev replied:
ok, I'll fix it

geomer198 commented on hr_timesheet_purchase_order/models/hr_timesheet_recurrence.py:
```diff
@@ -0,0 +1,351 @@
+
+MONTHS = {
+    "january": 31,
+    "february": 28,
```
On the form view I saw 31 days in select field.

dessanhemrayev replied:
This is a dictionary for counting the following dates

geomer198 commented on hr_timesheet_purchase_order/models/hr_employee_base.py:
```diff
@@ -8,3 +12,237 @@ class HrEmployeeBase(models.AbstractModel):
+    sat = fields.Boolean(string="Sat", compute="_compute_repeat", readonly=False)
+    sun = fields.Boolean(string="Sun", compute="_compute_repeat", readonly=False)
+
+    repeat_day = fields.Selection(
```
I think better use Integer field with min max options and add counter on the view.

dessanhemrayev replied:
okay

geomer198 commented on hr_timesheet_purchase_order/models/hr_employee_base.py:
```diff
@@ -8,3 +12,237 @@ class HrEmployeeBase(models.AbstractModel):
+    @api.model
+    def create(self, vals):
+        rec_fields = vals.keys() & self._get_recurrence_fields()
+        if rec_fields and vals.get("is_auto_po_generate") is True:
```
For what you use `is` in condition? You could just use `... and vals.get("is_auto_po_generate"):`

dessanhemrayev replied:
ok,i'll fix it

dessanhemrayev replied:
ok)

geomer198 commented on hr_timesheet_purchase_order/models/hr_employee_base.py:
```diff
@@ -8,3 +12,237 @@ class HrEmployeeBase(models.AbstractModel):
+                and (item.repeat_unit == "month" and item.repeat_on_month == "date")
+                or (item.repeat_unit == "year" and item.repeat_on_year == "date")
+            )
+            item.repeat_show_week = (
```
This compute value is the same at before, maybe use lambda function or separate for remove code duplication.

geomer198 commented on hr_timesheet_purchase_order/models/hr_employee_base.py:
```diff
@@ -8,3 +12,237 @@ class HrEmployeeBase(models.AbstractModel):
+                and (item.repeat_unit == "month" and item.repeat_on_month == "day")
+                or (item.repeat_unit == "year" and item.repeat_on_year == "day")
+            )
+            item.repeat_show_dow = (
```
Here too but optional.

geomer198 commented on hr_timesheet_purchase_order/models/hr_employee_base.py:
```diff
@@ -8,3 +12,237 @@ class HrEmployeeBase(models.AbstractModel):
+                if employee.recurrence_id:
+                    employee[f] = employee.recurrence_id[f]
+                else:
+                    if employee.is_auto_po_generate:
```
This condition you coulf write in one line:
```employee[f] = default.get(f) if employee.is_auto_po_generate else False```

dessanhemrayev replied:
Thanks, I'll fix it

geomer198 commented on hr_timesheet_purchase_order/models/hr_timesheet_recurrence.py:
```diff
@@ -0,0 +1,351 @@
+    sat = fields.Boolean(string="Sat")
+    sun = fields.Boolean(string="Sun")
+
+    repeat_day = fields.Selection([(str(i), str(i)) for i in range(1, 32)])
```
Here too better use integer field with borders

dessanhemrayev replied:
okay,I'll fix it

geomer198 commented on hr_timesheet_purchase_order/models/hr_timesheet_recurrence.py:
```diff
@@ -0,0 +1,351 @@
+    @api.constrains("repeat_unit", "mon", "tue", "wed", "thu", "fri", "sat", "sun")
+    def _check_recurrence_days(self):
+        for timesheet in self.filtered(lambda p: p.repeat_unit == "week"):
+            if not any(
```
Here you could use like this:
`if not any([getattr(timesheet, attr_name) for attr_name in DAYS]):`

dessanhemrayev replied:
okay, I'll fix it

geomer198 commented on hr_timesheet_purchase_order/models/hr_timesheet_recurrence.py:
```diff
@@ -0,0 +1,351 @@
+        "repeat_unit", "repeat_on_month", "repeat_day", "repeat_type", "repeat_until"
+    )
+    def _check_repeat_until_month(self):
+        if self.filtered(
```
This filter is too complex and doesn't have tests, please add test case on this behavior.

geomer198 commented on hr_timesheet_purchase_order/models/hr_timesheet_recurrence.py:
```diff
@@ -0,0 +1,351 @@
+        return [DAYS.get(self.repeat_weekday)(n)]
+
+    @api.model
+    def _get_next_recurring_dates(
```
This method is too complex, could you split functional on small methods

dessanhemrayev replied:
ok)

geomer198 reviewed (commented):
1) I can't understand for what you use compute method in all fields.
2) Readme need added.
3) For functions  need add comments and docstrings.
4) Test coverage not cover complex methods need fixed it.

geomer198 commented on hr_timesheet_purchase_order/models/hr_timesheet_recurrence.py:
```diff
@@ -106,7 +106,7 @@ class HRTimeSheetRecurrence(models.Model):
     sun = fields.Boolean(string="Sun")
 
-    repeat_day = fields.Selection([(str(i), str(i)) for i in range(1, 32)])
+    repeat_day = fields.Integer()
```
I think for this field need validation too.

dessanhemrayev replied:
ok, I'll add validation for this field

geomer198 reviewed (commented):
Please check comment

geomer198 reviewed (approved):
LGTM!

dessanhemrayev committed:
[FIX] hr_timesheet_purchase_order: fields for recurrency moved to billing_partner

dessanhemrayev committed:
[FIX] hr_timesheet_purchase_order: when create PO doesn't work methods onchange

dessanhemrayev committed:
[FIX] hr_timesheet_purchase_order: fix minor bugs

Change field label
Add validation for the setting product and add test for him
Add groups to the view

dessanhemrayev committed:
[FIX] hr_timesheet_purchase_order: change product description in PO

francesco-ooops commented on hr_timesheet_purchase_order/readme/USAGE.rst:
```diff
@@ -1,5 +1,7 @@
-* Click the *Create Purchase Order* in the  *Timesheet Sheet* form to create a new RFQ
+Go to Employees app > select an employee > go to HR Settings tab and enable the "Generate POs from timesheet sheet" checkbox
+Select the Billing partner which will be the vendor in the created POs
+By enabling "Automatic PO generation" user can set the recurrence of PO generation and whether the RFQ report should be sent automatically after creation
```
```suggestion
By enabling "Automatic PO generation from timesheet sheets" in Partner > Sales & Purchase tab, user can set the recurrence of PO generation and whether the RFQ report should be sent automatically after creation.
```

dessanhemrayev committed:
Update hr_timesheet_purchase_order/readme/USAGE.rst

Co-authored-by: Francesco Foresti <55489873+francesco-ooops@users.noreply.github.com>

dessanhemrayev committed:
[FIX] hr_timesheet_purchase_order: fix product description in PO

francesco-ooops reviewed (approved):
Functional ok!

geomer198 commented on hr_timesheet_purchase_order/models/hr_timesheet_recurrence.py:
```diff
@@ -0,0 +1,362 @@
+
+    @api.constrains("repeat_day", "repeat_month")
+    def _check_repeat_day_or_month(self):
+        if 0 > self.repeat_day or self.repeat_day > 31:
```
Not all months have 31 days. Please add condition for this.

dessanhemrayev replied:
okay, I'll fix it

geomer198 commented on hr_timesheet_purchase_order/models/hr_timesheet_recurrence.py:
```diff
@@ -0,0 +1,362 @@
+        return [DAYS.get(self.repeat_weekday)(n)]
+
+    @api.model
+    def _get_next_recurring_dates(
```
This method have so many arguments. Please use dict or `recurrence` record for this. For example:
```py
def _get_next_recurring_dates(self, repeat_struct):
```
`repeat_struct` is a dict or recurrence record.
This solution is easier and compact.

dessanhemrayev replied:
Thanks, I'll fix it

geomer198 commented on hr_timesheet_purchase_order/models/hr_timesheet_recurrence.py:
```diff
@@ -0,0 +1,362 @@
+        repeat_week,
+        repeat_month,
+        **kwargs
+    ):
```
Please add docstring for arguments.

dessanhemrayev replied:
okay, I'll add it

geomer198 commented on hr_timesheet_purchase_order/models/hr_timesheet_recurrence.py:
```diff
@@ -0,0 +1,362 @@
+}
+
+
+class HRTimeSheetRecurrence(models.Model):
```
Please add docstring for each method with describe arguments.

dessanhemrayev replied:
okay, I'll add it

geomer198 commented on hr_timesheet_purchase_order/models/res_partner.py:
```diff
@@ -0,0 +1,252 @@
+        readonly=False,
+    )
+    repeat_month = fields.Selection(
+        [
```
The same selection in prev model, would you create variable for this selection and use it in fields?

dessanhemrayev replied:
I can't do this because all these fields are calculated or modified by the user

geomer198 replied:
Selection argument is static data. Just create variable `MONTH_SELECTION`
```py
MONTH_SELECTION = [
("january", "January"),
("february", "February"),
....
]

...

repeat_month = fields.Selection(selection=MONTH_SELECTION)
```
The same for week day field

geomer198 commented on hr_timesheet_purchase_order/models/res_partner.py:
```diff
@@ -0,0 +1,252 @@
+    )
+    repeat_weekday = fields.Selection(
+        [
+            ("mon", "Monday"),
```
The same selection in prev model, would you create variable for this selection and use it in fields?

dessanhemrayev replied:
I can't do this because all these fields are calculated or modified by the user

geomer198 commented on hr_timesheet_purchase_order/models/res_partner.py:
```diff
@@ -0,0 +1,252 @@
+        if self.repeat_month == "february" and self.repeat_day > 29:
+            self.repeat_day = 28
+
+    repeat_week = fields.Selection(
```
The same selection in prev model, would you create variable for this selection and use it in fields?

dessanhemrayev replied:
I can't do this because all these fields are calculated or modified by the user

geomer198 commented on hr_timesheet_purchase_order/tests/test_create_not_product.py:
```diff
@@ -0,0 +1,62 @@
+from .common_po_recurrence import TestTimesheetPOrecurrenceCommon
+
+
+class TestTimesheetPORecurrenceNotProduct(TestTimesheetPOrecurrenceCommon):
```
Only one test on the all model? Would you please upped test coverage and add docstrings and assertions messages.

dessanhemrayev replied:
I understand that it is unnecessary to use an entire class for one test, but this requires the test to cover other functions of this module. And what kind of testing options can there be when the most important thing of this module is not selected, what is it intended for?)

dessanhemrayev replied:
I'll add docstrings and assertions messages.

geomer198 commented on hr_timesheet_purchase_order/tests/test_create_timesheet_po_recurrence.py:
```diff
@@ -0,0 +1,622 @@
+from .common_po_recurrence import TestTimesheetPOrecurrenceCommon
+
+
+class TestTimesheetPOrecurrence(TestTimesheetPOrecurrenceCommon):
```
Would you add docstrings and assertions messages

dessanhemrayev replied:
okay, I'll add it

geomer198 reviewed (changes requested):
Please fix your code.

dessanhemrayev committed:
[FIX] hr_timesheet_purchase_order: add maintainers

dessanhemrayev committed:
[FIX] hr_timesheet_purchase_order: refactoring of the code

alexey-pelykh commented on hr_timesheet_purchase_order/models/hr_timesheet_recurrence.py:
```diff
@@ -0,0 +1,399 @@
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+MONTHS = {
```
This hardcoded part should be used from corresponding library, `calendar.monthrange`

dessanhemrayev replied:
Thanks, for feedback, I'll fix it

alexey-pelykh commented on hr_timesheet_purchase_order/models/hr_timesheet_recurrence.py:
```diff
@@ -0,0 +1,399 @@
+    )
+    repeat_weekday = fields.Selection(
+        [
+            ("mon", "Monday"),
```
```
import locale

locale.nl_langinfo(locale.DAY_1)
```

don't multiply static text without need.

alexey-pelykh reviewed (changes requested):
Can't agree with amount of hardcoded calendar info

dessanhemrayev committed:
[FIX] hr_timesheet_purchase_order: Placed in a separate file constant values

geomer198 reviewed (approved):
LGTM!

dessanhemrayev committed:
[FIX] hr_timesheet_purchase_order: remove cetmixgitdrone from maintainers

francesco-ooops commented:
@alexey-pelykh shall we merge? thanks!

alexey-pelykh commented:
/ocabot merge minor
